OpenRA.Mods.Common/Traits/TargetableUnit.cs

```diff
@@ -52,5 +63,26 @@ public virtual IEnumerable<WPos> TargetablePositions(Actor self)
 		}
 
 		public bool RequiresForceFire { get { return info.RequiresForceFire; } }
+
+		public IEnumerable<string> UpgradeTypes	{ get { return info.UpgradeTypes; } }
+
+		public bool AcceptsUpgradeLevel(Actor self, string type, int level)
+		{
+			return level <= 1;
+		}
+
+		public void UpgradeLevelChanged(Actor self, string type, int oldLevel, int newLevel)
+		{
+			if (newLevel > 0)
+			{
+				if (info.UpgradedTargetTypes.Contains("!"))
+					targetTypes = info.UpgradedTargetTypes.Where(utt => !utt.StartsWith("!")).ToArray();
+				else
+					targetTypes = targetTypes.Except(info.UpgradedTargetTypes.Where(utt => utt.StartsWith("!")).Select(utt => utt.Substring(1)))
```

Comment:
Sorry if I missunderstood something here, but aren't you checking for `info.UpgradedTargetTypes.Contains("!")` above, so `info.UpgradedTargetTypes.Where(utt => utt.StartsWith("!")).Select(...` is redundant here?
(As something that doesn't contain a "!" shouldn't start with it :p)